### PR TITLE
dvc: refactor package versioning logic

### LINKED ISSUE #1824: Wrong version number
When installing DVC using pip (`pip install dvc`), the version information supplied by dvc does not match that reported by pip:

Specifically, pip reports that 0.34.0 is installed, but `dvc --version` returns:

0.29.0+220b4b.mod

This is also confirmed by running the following command:

```
python -c 'import dvc; print(dvc.VERSION)'
0.29.0+220b4b.mod
```

```
python -c 'import dvc; print(dvc.VERSION_BASE)'
0.34.0
```

**Please provide information about your setup**

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename:	bionic
```

### PR DESCRIPTION
**Pull request by mroutis**
- Use `version.py` to compute a dynamic version stored in `__version__`.
  Version information was scattered across the source code
  (`VERSION_BASE`, `VERSION`, `__version__`, and `version.py`).

- Read `__version__` from `setup.py` and **pin** it using custom
  `build_py` command class.
  This way, we can maintain consistancy between the build version
  and the runtime one.
  Also, this gives us the ability to run `setup.py` without depending on
  on `dvc` itself.

Close #1824

---

[asciicast](https://asciinema.org/a/239169)

### CODE REVIEW

**Comment by efiop** [comment]
@mroutis Try this with your patch:
```
pip uninstall -y dvc; rm -rf build dist && ./scripts/build_package.sh && pip install dist/dvc-*.tar.gz
```
you'll get
```
$ dvc --version
0.34.2
```
and
```
$ pip freeze | grep dvc
dvc==0.34.2
```
And you will get the same if you try to install from github directly with `pip install git+https://github.com/mroutis/dvc@fix-1824`.
In comparison, current behaviour would be:
```
$ dvc --version
0.34.2+291fa0
```
and
```
$ pip freeze | grep dvc
dvc==0.34.2+291fa0
```
This is because we include auto-generated `version.py` when we build our package. Your patch will only work if you are running dvc directly from git directly. And the current approach works in all cases, except that it doesn't compare what it has read from `version.py` with VERSION_BASE that it has to verify that `version.py` is bad and needs to be re-generated, which causes https://github.com/iterative/dvc/issues/1824 . Adding that logic would fix #1824 .

**Comment by mroutis** [comment]
@efiop , great catch!

However, the current implementation also lacks some edge cases:
```bash
# Make sure you don't have any environment active!
virtualenv -p python2.7 .venv
source .venv/bin/activate
pip install git+https://github.com/iterative/dvc
dvc --version
# 0.32.4

pip uninstall -y dvc
# Version 0.34.0
pip install git+https://github.com/iterative/dvc@a41705a757650f0841b6b331d61afb726ff278cb
dvc --version
# 0.34.0
```
Master:


From this patch:


----

@efiop , please take a look into my latest commit, it uses a more _sound_ approach, by computing the version with git information from `setup.py`, the only downside would be for people running development versions, it would execute `git rev-parse HEAD` and `git status` on every command.

**Comment by efiop** [comment]
@mroutis 

```
$ git clone https://github.com/mroutis/dvc@fix-1824
$ cd dvc
$ pip uninstall -y dvc; rm -rf build dist && ./scripts/build_package.sh && pip install dist/dvc-*.whl
$ dvc --version
0.34.2
```
still doesn't work. Also, calling `git status` each time is a problem.

**Comment by efiop** [comment]
@mroutis 

> However, the current implementation also lacks some edge cases:

It does if you have gitpython installed :) But that is a great point! We could totally try to call `git` directly there as you do in your current patch. But I would still stick with the current `version.py` auto-generation logic overall.

**Comment by mroutis** [comment]
> still doesn't work.

Ohh! It is because even if `setup.py` has the correct version, the project doesn't have access anymore to the build version (no `.git`), that's why you were using the previous `version.py` as a cache!



I'll push one last commit, @efiop, promise!

(By the way, thanks for bearing with me all the way to this point :sweat_smile: )

**Comment by mroutis** [comment]
@efiop , I added a gif with some testing.

That last commit was my _end game_, I will not try to pursuit other solutions.
If you like the one in the PR let's move with it, if not, I can rollback to our previous versioning and submit a small patch with the comparison you suggested since the beginning of this task :sweat_smile:

File: setup.py [hunk 1]
```diff
@@ -1,5 +1,37 @@
...
+#
+# References:
+#   - https://docs.python.org/3.7/distutils/extending.html
+#   - https://github.com/python/mypy
+class build_py(_build_py):
+    def pin_version(self):
+        path = os.path.join(self.build_lib, "dvc")
+        self.mkpath(path)
+        with open(os.path.join(path, "version.py"), "w") as stream:
+            stream.write('__version__ = "{}"\n'.format(version))
```
Comment (efiop) [inline setup.py]: Let's at least put a header comment that this file was autogenerated by `setup.py`. Similar to how we have a comment about it in the current version logic.

**Comment by efiop** [comment]
@mroutis Looks really great! Loving how it looks! :fire:

File: setup.py (hunk 1 above)
Comment (mroutis) [reply setup.py]: :+1:

**Pull request merged**